### PR TITLE
Add Athenz authentication plugin

### PR DESCRIPTION
**Pull request by nkurihar**
### Motivation

[Athenz](https://github.com/yahoo/athenz) is a role-based authorization system which is recently open-sourced by Yahoo! inc.
We want to use Athenz as an authentication mechanism on pulsar.

### Modifications

Add Athenz authentication plugin classes:
- `AuthenticationProviderAthenz` (for Broker)
- `AuthenticationAthenz` (for Client)
- `AuthenticationDataAthenz` (for Client)

and add `athenzDomainNames` congifuration setting to broker.conf / standalone.conf

### Result

Be able to use Athenz authentication on pulsar.

### TODO

Add documents about Athenz authentication on pulsar.

### CODE REVIEW

File: pulsar-broker-common/src/main/java/com/yahoo/pulsar/broker/authentication/AuthenticationProviderAthenz.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        log.info("Supported domain names for athenz: {}", domainNameList);
+    }
+
+    @Override
+    public String getAuthMethodName() {
+        return "athenz";
+    }
+
+    @Override
+    public String authenticate(AuthenticationDataSource authData) throws AuthenticationException {
```
Comment (merlimat) [inline pulsar-broker-common/src/main/java/com/yahoo/pulsar/broker/authentication/AuthenticationProviderAthenz.java]: How does this auth provider interacts with the authorization policies in Pulsar? Do one needs to "grant permission" on the namespace in Pulsar or everything is handled in Athenz?

File: pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/auth/AuthenticationAthenz.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+import com.yahoo.athenz.auth.ServiceIdentityProvider;
+import com.yahoo.athenz.auth.impl.SimpleServiceIdentityProvider;
+import com.yahoo.athenz.auth.util.Crypto;
+import com.yahoo.pulsar.client.api.Authentication;
+import com.yahoo.pulsar.client.api.AuthenticationDataProvider;
+import com.yahoo.pulsar.client.api.PulsarClientException;
+import com.yahoo.pulsar.client.api.PulsarClientException.GettingAuthenticationDataException;
+
+/**
+ * Created by vmahedia on 10/3/15.
```
Comment (merlimat) [inline pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/auth/AuthenticationAthenz.java]: Classes authors should be removed

File: pulsar-broker-common/pom.xml [hunk 3]
```diff
@@ -51,5 +51,11 @@
       <groupId>javax.servlet</groupId>
       <artifactId>javax.servlet-api</artifactId>
     </dependency>
+
+    <dependency>
```
Comment (nkurihar) [inline pulsar-broker-common/pom.xml]: I' ll move athenz plugin codes to another module directory(like "pulsar-auth-athenz"), thanks.

File: pulsar-broker-common/src/main/java/com/yahoo/pulsar/broker/authentication/AuthenticationProviderAthenz.java (hunk 1 above)
Comment (nkurihar) [inline pulsar-broker-common/src/main/java/com/yahoo/pulsar/broker/authentication/AuthenticationProviderAthenz.java]: > everything is handled in Athenz?
No.
In this PR, we use Athenz for only authentication,  not authorization.
So users should do "grant permission" on pulsar admin API.

I think to use Athenz for both authentication and authorization is possible,
maybe it needs abstraction of authorization logic on pulsar, 
and should be discussed as another issue.

**Comment by nkurihar** [comment]
> end-to-end test using Athenz

Do you mean unit tests ?
OK, I think the test is necessary too, I'll try it.

**Comment by merlimat** [comment]
@nkurihar The change looks good to me. If it's feasible, it would be good to have a unit test that spawns up an Athenz service and verify the authentication end-to-end. 

Also, it would be good to have a new page in the documentation to show how to enable and configure the authentication with Athenz 😄 . But that can go in a different PR.

**Comment by nkurihar** [comment]
@merlimat
Thank you for your reviewing.

> new page in the documentation to show how to enable and configure the authentication with Athenz

I'll add documents as another PR.

> have a unit test that spawns up an Athenz service and verify the authentication end-to-end

I'll investigate and if feasible, add the end-to-end unit test as another PR.

**Pull request merged**